In [2]:
import pandas as pd
import re
import numpy as np

In [137]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
adverse=pd.read_csv('adverse_reactions.csv')
treatment_cut=pd.read_csv('treatments_cut.csv')

In [71]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)
print(treatment_cut.shape)

(503, 14)
(280, 7)
(34, 3)
(70, 7)


In [72]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [73]:
patients.tail()

patient_id assigned_sex given_name      surname  \
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
498       2530 Victoria Court   Milton Mills        ME    3852.0   
499  494 Clarksburg Park Road         Sedona        AZ   86341.0   
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country                                          contact  \
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3

In [74]:
patients[patients.duplicated()]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, contact, birthdate, weight, height, bmi]
Index: []

In [75]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [76]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [77]:
patients_copy[patients_copy['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [78]:
patients.sample(10)

patient_id assigned_sex  given_name    surname               address  \
35           36       female      Kamila   Pecinová  3558 Longview Avenue   
199         200         male      Zdeněk      Synek  3818 Strother Street   
342         343       female    Mathilde   Nørgaard   846 Copperhead Road   
472         473       female       Julia     Werner   325 Eastland Avenue   
409         410       female      Albika  Ibragimov    4945 Jarvis Street   
96           97       female       Nasim      Salib     3161 Fantages Way   
213         214       female  Onyemaechi  Onwughara     685 Duncan Avenue   
401         402       female      Kasumi       Hori    1116 Dog Hill Lane   
357         358       female    Chiemela  Tobeolisa    2881 Harter Street   
463         464       female       Bouke     Glaser      3006 Maple Court   

              city     state  zip_code        country  \
35        New York  New York   10004.0  United States   
199           Dora        AL   35062.0  United States   
342       Hartford        CT    6182.0  United States   
472        Natchez        MS   39120.0  United States   
409        Buffalo  New York   14211.0  United States   
96    Presque Isle        ME    4769.0  United States   
213       New York        NY   10013.0  United States   
401       Lawrence        KS   66044.0  United States   
357  Bellefontaine        OH   43311.0  United States   
463     Owensville        MO   65066.0  United States   

                                          contact   birthdate  weight  height  \
35          718-501-0503KamilaPecinova@dayrep.com  12/23/1985   198.9      62   
199      ZdenekSynek@jourrapide.com1 205 648 4197   9/27/1950   168.3      70   
342  MathildeDNrgaard@superrito.com1 860 618 4560   1/21/1985   127.2      61   
472           601-446-6829JuliaWerner@teleworm.us   9/14/1938   155.3      61   
409     AlbikaIbragimov@superrito.com716-897-5050   2/29/1948   161.7      63   
96       207-768-0477NasimSumaiyaSalib@einrot.com   7/30/1971   211.4      65   
213    917-622-9142OnyemaechiOnwughara@einrot.com    3/8/1989   131.1      69   
401         KasumiHori@gustr.com+1 (785) 748-8181   3/30/1959   180.6      62   
357       937-592-9410ChiemelaTobeolisa@rhyta.com    6/7/1965   187.9      69   
463          BoukeGlaser@einrot.com1 573 437 7334  10/21/1996   142.1      66   

      bmi  
35   36.4  
199  24.1  
342  24.0  
472  29.3  
409  28.6  
96   35.2  
213  19.4  
401  33.0  
357  27.7  
463  22.9

In [79]:
patients[patients['contact'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [80]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [81]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [82]:
treatment[treatment.duplicated()]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [83]:
treatment[treatment['given_name']=='joseph']

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
10      joseph     day  29u - 36u       -          7.7       7.19   
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
10            NaN  
136           NaN

In [84]:
treatment[treatment['hba1c_start'].isnull()]

Empty DataFrame
Columns: [given_name, surname, auralin, novodra, hba1c_start, hba1c_end, hba1c_change]
Index: []

In [85]:
treatment[treatment['hba1c_end'].isnull()]

Empty DataFrame
Columns: [given_name, surname, auralin, novodra, hba1c_start, hba1c_end, hba1c_change]
Index: []

In [86]:
treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [87]:
treatment.tail()

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36

In [88]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [89]:
treatment_cut.describe()

hba1c_start  hba1c_end  hba1c_change
count    70.000000  70.000000     42.000000
mean      7.838000   7.443143      0.518810
std       0.423007   0.418706      0.270719
min       7.510000   7.020000      0.280000
25%       7.640000   7.232500      0.340000
50%       7.730000   7.345000      0.370000
75%       7.860000   7.467500      0.907500
max       9.910000   9.460000      0.970000

In [90]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [91]:
adverse.tail()

given_name   surname adverse_reaction
29       alex  crawford     hypoglycemia
30     monika    lončar     hypoglycemia
31     steven       roy         headache
32    cecilie    nilsen     hypoglycemia
33  krisztina    magyar     hypoglycemia

In [92]:
adverse[adverse.duplicated()]

Empty DataFrame
Columns: [given_name, surname, adverse_reaction]
Index: []

In [93]:
adverse[adverse['adverse_reaction'].isnull()]

Empty DataFrame
Columns: [given_name, surname, adverse_reaction]
Index: []

In [94]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 944.0+ bytes


In [95]:
adverse.describe()

given_name  surname adverse_reaction
count          34       34               34
unique         34       33                6
top        jia li  johnson     hypoglycemia
freq            1        2               19

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format V]
- Certain zip codes of 4 digits [V]  
- For the state col sometimes full names are use and other times short forms are used [IC]
- Missing values for contact details of 12 patients [C]
- Incorrect data types for assigned_sex and birthdate cols [?]  
- Erroneous data for weights and heights [Accuracy]
- Inconsistent way of storing phone numbers in contact col [IC]  
- Duplicate and not useful data in dataset [IC]

#### `Treatments`

- hba1c_change col has missing values [C]  
- Novodra and Auralin cols has u attached to numbers [V]  
- Only 280 rows are there, instead of 350 [C] 
- Name and surname col is in lower case [IC]   
- Incorrect data type for auralin and novadra cols [V]
- hba1c_change 9 instead of 4 [A] 
- use of dashes instead of Nan in aurolin and novodra cols [IC]  
- duplicated row [IC] 

#### `Adverse Reactions`

- Name and surname col is in lower case [IC]  

### Tidiness

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Phone and email together in one col i.e. contact [IC]

#### `Treatments`

- 3 variables stored in auralin and novodra cols [IC]

#### `Adverse Reactions`

- No point of using this table [IC]


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [157]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()
treatment_cut_copy=treatment_cut.copy()

## a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

####  Define: Missing values for contact details of 12 patients

`Solution`: There is no way we can retrieve lost contact information nor a way to infer the details, so we don't have any choice other than to reject it.  

####  Define: Only 280 rows are there, instead of 350
`Solution`: We have another DataFrame treatment_cut which contains the missing 70 columns. We will have to append it this dataframe.

#### Code

In [119]:
treatment_copy=pd.concat([treatment_copy,treatment_cut_copy],ignore_index=True)

#### Test

In [120]:
treatment_copy

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika    jindrová  41u - 48u          -         7.63       7.20   
1       elliot  richardson          -  40u - 45u         7.56       7.09   
2     yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3         skye  gormanston  33u - 36u          -         7.97       7.62   
4       alissa      montez          -  33u - 29u         7.78       7.46   
..         ...         ...        ...        ...          ...        ...   
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN  

[350 rows x 7 columns]

####  Define: hba1c_change col has missing values
`Solution`: Subtract hba1c_start from hba1c_end to get hba1c_change

#### Code

In [123]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

#### Test

In [109]:
treatment_copy.info()

treatment_copy.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


hba1c_start   hba1c_end  hba1c_change
count   350.000000  350.000000    350.000000
mean      7.956343    7.560057      0.396286
std       0.545328    0.545456      0.059283
min       7.500000    7.010000      0.200000
25%       7.650000    7.270000      0.350000
50%       7.785000    7.400000      0.390000
75%       7.950000    7.557500      0.440000
max       9.950000    9.580000      0.540000

## b. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

#### Define:  Novodra and Aurolin occupies separate columns with 'u' suffux
`Solution`: Create a new column named treatment for the type of medicine prescribed using the melt function

#### Code

In [121]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatment_copy=treatment_copy[treatment_copy['dose']!='-']
treatment_copy['lower_dosage'],treatment_copy['upper_dosage']=treatment_copy['dose'].str.split('-').str
treatment_copy.drop(columns={'dose'},inplace=True)
treatment_copy['lower_dosage']=pd.to_numeric(treatment_copy.lower_dosage.str.extract(pat = '([0-9].)',expand=False))
treatment_copy['upper_dosage']=pd.to_numeric(treatment_copy.upper_dosage.str.extract(pat = '([0-9].)',expand=False))
treatment_copy['treatment']=treatment_copy['treatment'].astype('category')

#### Test

In [135]:
treatment_copy.sample(5)

given_name   surname  hba1c_start  hba1c_end  hba1c_change  \
56   zikoranaudodimma  chinedum         8.19       7.83          0.36   
219           ananías  enríquez         7.58       7.07          0.51   
300              kang       mai         7.78       7.45          0.33   
162              anja   mueller         7.58       7.29          0.29   
101           maximus    henzen         7.79       7.34          0.45   

    treatment  lower_dosage  upper_dosage           adverse_reaction  
56    auralin            41            51                        NaN  
219   novodra            44            45                        NaN  
300   novodra            39            36  injection site discomfort  
162   auralin            43            56                        NaN  
101   auralin            49            55                        NaN

#### Define: No Point in making new table (Adverse) when a single column can be used in Treatment to represent the same data
`Solution`: Merge the adverse table to the treatment table

#### Code

In [131]:
treatment_copy=treatment_copy.merge(adverse_copy,on=['given_name','surname'],how='outer')

#### Test

In [132]:
treatment_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
given_name          350 non-null object
surname             350 non-null object
hba1c_start         350 non-null float64
hba1c_end           350 non-null float64
hba1c_change        350 non-null float64
treatment           350 non-null object
lower_dosage        350 non-null int64
upper_dosage        350 non-null int64
adverse_reaction    35 non-null object
dtypes: float64(3), int64(2), object(4)
memory usage: 27.3+ KB


#### Define: Inconsistent way of storing phone numbers in contact col
`Solution`: Create separate columns for email_id and phone numbers and minimise incosistency in their format

#### Code

In [158]:
patients_copy['email_id']=patients_copy.contact.str.extract(pat = '([a-z,A-Z,@,.]+)',expand=False)

def phone_num(a):
    try:
        num='-'.join(re.findall(r'\d{3}',a)[:-1]+re.findall(r'\d{4}',a))
        if num=='':
            return(np.nan)
        return(num)
    except:
        return(np.nan)
    #print(type(a))
    
patients_copy['phone_number']=patients_copy['contact'].apply(phone_num)

patients_copy.drop(columns={'contact'},inplace=True)

#### Test

In [159]:
patients_copy.sample(5)

patient_id assigned_sex given_name     surname  \
269         270       female     Flavia  Fiorentino   
192         193       female       Jade      Parker   
205         206         male      Breno     Correia   
85           86         male    Burrell     Mathieu   
441         442         male       Luis       Silva   

                        address         city     state  zip_code  \
269                         NaN          NaN       NaN       NaN   
192  2957 Feathers Hooves Drive  Garden City        NY   11530.0   
205            3539 Bottom Lane    Tonawanda  New York   14150.0   
85            3816 Grand Avenue    Kissimmee   Florida   34746.0   
441           959 Heritage Road  San Joaquin        CA   93660.0   

           country  birthdate  weight  height   bmi  \
269            NaN  10/9/1937   175.2      61  33.1   
192  United States  6/19/1927   188.8      63  33.4   
205  United States  7/11/1974   180.6      67  28.3   
85   United States  5/16/1936   203.5      71  28.4   
441  United States  11/2/1946   239.1      71  33.3   

                            email_id  phone_number  
269                              NaN           NaN  
192         JadeParker@superrito.com  631-704-6487  
205   BrenoLimaCorreia@superrito.com  716-743-5884  
85        BurrellMathieu@teleworm.us  407-662-9389  
441  LuisRibeiroSilva@jourrapide.com  559-693-6779

## c. Validity Issues
- Zip code col is stored in decimal format
- Certain zip codes of 4 digits.

#### Define: Zip code col is stored in decimal format and some are of 4 digits instead of 5
`Solution`:As it impossible to deduce the wrong zip codes,ou option is to add zeros to the left to make it 5 digits

#### Code

In [165]:
patients_copy['zip_code']=patients_copy['zip_code'].astype('str')
patients_copy['zip_code']=patients_copy['zip_code'].str.replace('.0','').str.pad(width=5,side='left',fillchar='0')
#patients_copy['zip_code']=pd.to_numeric(patients_copy['zip_code'])

#### Test

In [164]:
patients_copy[patients_copy['zip_code'].isnull()]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email_id, phone_number]
Index: []

In [161]:
patients_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
email_id        491 non-null object
phone_number    491 non-null object
dtypes: float64(2), int64(2), object(11)
memory usage: 59.1+ KB


## d. Accuracy issues
- Erroneous data for weights and heights

### Define: Erroneous data for weight and height
#### Example

In [100]:
patients_copy[patients_copy['height']<55]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      27  26.1   

                 email  phone_number  
4  TimNeudorf@cuvox.de  334-515-7487

Height of 27 inches is just not acceptable. Now this might be a valid person but when mathematical algorithms are applied on this column, this value might give us erroneous results       
`Solution`: We are going to remove this row
#### Code

In [101]:
patients_copy.drop(index=4,inplace=True)

#### Test

In [102]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
5           6         male     Rafael    Costa         1140 Willis Avenue   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
5      Daytona Beach     Florida    32114  United States  8/31/1931   183.9   

   height   bmi                         email  phone_number  
0      66  19.6      ZoeWellish@superrito.com  951-719-9170  
1      66  19.2          PamelaSHill@cuvox.de  217-569-3204  
2      71  24.8          JaeMDebord@gustr.com  402-363-6804  
3      70  31.7     PhanBaLiem@jourrapide.com  732-636-8246  
5      70  26.4  RafaelCardosoCosta@gustr.com  386-334-5237

## e. Inconsistency Issues
- For the state col sometimes full names are use and other times short forms are used
- Duplicate and not useful data in dataset
- Name and surname col is in lower case

### Define: Name and surname col is in lower case
`Solution`: Capitalize them


#### Code

In [103]:
treatment_copy['given_name']=treatment_copy['given_name'].str.capitalize()
treatment_copy['surname']=treatment_copy['surname'].str.capitalize()

#### Test

In [104]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   Veronika    Jindrová         7.63       7.20          0.43   auralin   
1       Skye  Gormanston         7.97       7.62          0.35   auralin   
2     Sophia      Haugen         7.65       7.27          0.38   auralin   
3      Eddie      Archer         7.89       7.55          0.34   auralin   
4       Asia     Woźniak         7.76       7.37          0.39   auralin   

   start  end adverse_reaction  
0     41   48              NaN  
1     33   36              NaN  
2     37   42              NaN  
3     31   38              NaN  
4     30   36              NaN

### Define: Duplicate and not useful data in dataset

#### Code

In [105]:
# In treatment dataframe
treatment_copy[treatment_copy.duplicated(subset=['given_name','surname'])]

given_name surname  hba1c_start  hba1c_end  hba1c_change treatment  start  \
62     Joseph     Day          7.7       7.19          0.51   auralin     29   

    end adverse_reaction  
62   36     hypoglycemia

In [106]:
treatment_copy.drop_duplicates(inplace=True)

In [107]:
# In patients dataframe
patients_copy[patients_copy.duplicated(subset=['given_name','surname'])]

patient_id assigned_sex given_name surname          address      city  \
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country birthdate  weight  height   bmi  \
229    NY    12345  United States  1/1/1975   180.0      72  24.4   
237    NY    12345  United States  1/1/1975   180.0      72  24.4   
244    NY    12345  United States  1/1/1975   180.0      72  24.4   
251    NY    12345  United States  1/1/1975   180.0      72  24.4   
277    NY    12345  United States  1/1/1975   180.0      72  24.4   

                 email phone_number  
229  johndoe@email.com          NaN  
237  johndoe@email.com          NaN  
244  johndoe@email.com          NaN  
251  johndoe@email.com          NaN  
277  johndoe@email.com          NaN

In [108]:
patients_copy.drop_duplicates(subset=['given_name','surname'],inplace=True)

#### Test

In [109]:
treatment_copy.shape


(349, 9)

In [110]:
patients_copy.shape

(497, 15)

### Define: For the state col sometimes full names are use and other times short forms are used

#### Problem

In [111]:
patients_copy['state'].unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'Florida', 'NV', 'CA',
       'MO', 'New York', 'MI', 'TN', 'VA', 'AL', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

#### `Solution`: Replace California, Illinois, Nebraska and Florida to CA, IL, NE and FL resp.

#### Code


In [112]:
patients_copy['state'].replace({'California':'CA','Illinois':'IL','Nebraska':'NE','Florida':'FL'},inplace=True)

#### Test

In [113]:
patients_copy.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
5           6         male     Rafael    Costa         1140 Willis Avenue   

                city state zip_code        country  birthdate  weight  height  \
0  Rancho California    CA    92390  United States  7/10/1976   121.7      66   
1          Armstrong    IL    61812  United States   4/3/1967   118.8      66   
2               York    NE    68467  United States  2/19/1980   177.8      71   
3         Woodbridge    NJ    07095  United States  7/26/1951   220.9      70   
5      Daytona Beach    FL    32114  United States  8/31/1931   183.9      70   

    bmi                         email  phone_number  
0  19.6      ZoeWellish@superrito.com  951-719-9170  
1  19.2          PamelaSHill@cuvox.de  217-569-3204  
2  24.8          JaeMDebord@gustr.com  402-363-6804  
3  31.7     PhanBaLiem@jourrapide.com  732-636-8246  
5  26.4  RafaelCardosoCosta@gustr.com  386-334-5237

# Your cleaned data

In [114]:
patients=patients_copy.copy()
treatment=treatment_copy.copy()


In [115]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
5           6         male     Rafael    Costa         1140 Willis Avenue   

                city state zip_code        country  birthdate  weight  height  \
0  Rancho California    CA    92390  United States  7/10/1976   121.7      66   
1          Armstrong    IL    61812  United States   4/3/1967   118.8      66   
2               York    NE    68467  United States  2/19/1980   177.8      71   
3         Woodbridge    NJ    07095  United States  7/26/1951   220.9      70   
5      Daytona Beach    FL    32114  United States  8/31/1931   183.9      70   

    bmi                         email  phone_number  
0  19.6      ZoeWellish@superrito.com  951-719-9170  
1  19.2          PamelaSHill@cuvox.de  217-569-3204  
2  24.8          JaeMDebord@gustr.com  402-363-6804  
3  31.7     PhanBaLiem@jourrapide.com  732-636-8246  
5  26.4  RafaelCardosoCosta@gustr.com  386-334-5237

In [116]:
treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   Veronika    Jindrová         7.63       7.20          0.43   auralin   
1       Skye  Gormanston         7.97       7.62          0.35   auralin   
2     Sophia      Haugen         7.65       7.27          0.38   auralin   
3      Eddie      Archer         7.89       7.55          0.34   auralin   
4       Asia     Woźniak         7.76       7.37          0.39   auralin   

   start  end adverse_reaction  
0     41   48              NaN  
1     33   36              NaN  
2     37   42              NaN  
3     31   38              NaN  
4     30   36              NaN

### What is Assessing Data?

The data analysis process consists of the following steps

1. Asking Questions
2. Data Wrangling<br><br>
    a. Gathering Data
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data</b><br>
    c. Cleaning Data<br><br>
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results
    

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues